### PR TITLE
Add Tabular Diff for CSV files

### PR DESCRIPTION
**Pull request by KN4CK3R**
Implements request #14320 The rendering of csv files does match the diff style.

A simple change:
grafik

Changes in different areas (note the jumping line numbers):
grafik

A removed column is rendered without marking all rows as changed:
grafik

Toggle view:
toggle

### CODE REVIEW

File: routers/repo/compare.go [hunk 1]
```diff
@@ -65,6 +78,49 @@ func setImageCompareContext(ctx *context.Context, base *git.Commit, head *git.Co
...
+				return nil, err
+			}
+
+			reader, err := blob.DataAsync()
+			if err != nil {
+				return nil, err
+			}
+			defer reader.Close()
+
+			b, err := ioutil.ReadAll(reader)
```
Comment (zeripath) [inline routers/repo/compare.go:119]: This reader is unbound in size. There's no reason why that couldn't simply read in 100Mb straight in to memory.

**Review by zeripath — DISMISSED** [review]

File: services/gitdiff/csv.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+			cells[j] = &TableDiffCell{LeftCell: row[j], Type: celltype}
+		}
+		rows[i] = &TableDiffRow{RowIdx: i + 1, Cells: cells}
+	}
+
+	return []*TableDiffSection{{Rows: rows}}, nil
+}
+
+func createCsvDiff(diffFile *DiffFile, baseReader *csv.Reader, headReader *csv.Reader) ([]*TableDiffSection, error) {
+	a, err := baseReader.ReadAll()
```
Comment (zeripath) [inline services/gitdiff/csv.go]: This is also an unbound read.

File: services/gitdiff/csv.go [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := headReader.ReadAll()
```
Comment (zeripath) [inline services/gitdiff/csv.go]: And this.

File: routers/repo/compare.go (hunk 1 above)
Comment (KN4CK3R) [reply routers/repo/compare.go:119]: I have ignored this because this code doesn't run if the diff decides "file too large". Anyway the max file size is restricted to 512kb now like on GitHub.

**Comment by KN4CK3R** [comment]
File sizes are now restricted to 512kb.
grafik

This is a problem for the old csv markup renderer because there is no way to return errors from the rendering process and fall back to the normal text rendering. If the file is too large the content is now rendered without styling:
grafik
With a fallback to the normal rendering combined with an error message from the markup renderer it could be displayed like on GitHub:
grafik

Additionaly the csv parser errors are now propagated to the ui:
grafik

File: modules/base/csv.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (lunny) [inline modules/base/csv.go]: Why not keep these code still in package `modules/markup/csv/`?

Comment (KN4CK3R) [reply modules/base/csv.go]: I thought the csv reader creation logic shouldn't be coupled with the render part.

Comment (lunny) [reply modules/base/csv.go]: I'm OK with another package but put them in `base` package seems not suitable. Maybe we can create a new package named `modules/csv` or similar to do that.

Comment (KN4CK3R) [reply modules/base/csv.go]: I moved the code to a new csv package.

File: modules/csv/csv.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			bestScore = score
+			bestDelim = delim
+		}
+	}
+
+	return bestDelim
+}
+
+// scoreDelimiter uses a count & regularity metric to evaluate a delimiter against lines of CSV.
+func scoreDelimiter(lines []string, delim rune) float64 {
```
Comment (lunny) [inline modules/csv/csv.go:59]: How did you handle `"a,b"\t"c,b"`?

File: modules/csv/csv.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+// guessDelimiter scores the input CSV data against delimiters, and returns the best match.
+// Reads at most 10k bytes & 10 lines.
+func guessDelimiter(data []byte) rune {
+	maxLines := 10
+	maxBytes := util.Min(len(data), 1e4)
+	text := string(data[:maxBytes])
+	text = quoteRegexp.ReplaceAllLiteralString(text, "")
+	lines := strings.SplitN(text, "\n", maxLines+1)
```
Comment (lunny) [inline modules/csv/csv.go:41]: `"a\nb","c"`?

File: modules/csv/csv.go (hunk 5 above)
Comment (lunny) [reply modules/csv/csv.go:59]: This is not for this PR since it's only a refactor.

File: modules/csv/csv.go (hunk 6 above)
Comment (lunny) [reply modules/csv/csv.go:41]: This is not for this PR since it's only a refactor.

File: modules/csv/csv.go (hunk 5 above)
Comment (KN4CK3R) [reply modules/csv/csv.go:59]: I didn't change this code but first the regex deletes all quoted text and then the delimiter `\t` is found.
Original code: https://github.com/go-gitea/gitea/blob/0a9a484e1e2c082f7ebd747837e9c9557b3bacac/modules/markup/csv/csv.go#L67-L113

File: modules/csv/csv.go (hunk 6 above)
Comment (KN4CK3R) [reply modules/csv/csv.go:41]: Same as [above](https://github.com/go-gitea/gitea/pull/14661#discussion_r576360598).
grafik

File: custom/conf/app.example.ini [hunk 7]
```diff
@@ -248,6 +248,10 @@ EVENT_SOURCE_UPDATE_TIME = 10s
 ; Whether to render SVG files as images.  If SVG rendering is disabled, SVG files are displayed as text and cannot be embedded in markdown files as images.
 ENABLE_RENDER = true
 
+[ui.csv]
+; Maximum allowed file size in bytes to render CSV files as table. (Set to 0 for no limit).
+MAX_FILE_SIZE = 524288
```
Comment (bagasme) [inline custom/conf/app.example.ini:253]: Why did you set 512 KB as default maximum?

Comment (KN4CK3R) [reply custom/conf/app.example.ini:253]: I just adopted the value from Github.

> Exceeding the file size. Our rendering only works for files up to 512KB. Anything bigger than that slows down the browser. [Source](https://docs.github.com/en/github/managing-files-in-a-repository/rendering-csv-and-tsv-data)

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 8]
```diff
@@ -758,7 +762,7 @@ NB: You must have `DISABLE_ROUTER_LOG` set to `false` for this option to take ef
 
 - `PATH`: **""**: The path of git executable. If empty, Gitea searches through the PATH environment.
 - `DISABLE_DIFF_HIGHLIGHT`: **false**: Disables highlight of added and removed changes.
-- `MAX_GIT_DIFF_LINES`: **100**: Max number of lines allowed of a single file in diff view.
+- `MAX_GIT_DIFF_LINES`: **1000**: Max number of lines allowed of a single file in diff view.
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Is this really related?

Comment (KN4CK3R) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]: No, should this "fix" be another PR?

Comment (lafriks) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]: Yes, please

Comment (KN4CK3R) [reply docs/content/doc/advanced/config-cheat-sheet.en-us.md]: done #14949

File: web_src/less/_repository.less [hunk 9]
```diff
@@ -1444,6 +1444,66 @@
...
+    tr {
+      border-top: 0;
+    }
+
+    td,
+    th {
+      padding: 5px !important;
+      overflow: hidden;
+      font-size: 12px;
+      line-height: 1;
```
Comment (kdumontnu) [inline web_src/less/_repository.less]: I would use something a little higher, like 1.2. Gets a little squished with some fonts.

image

**Review by kdumontnu — CHANGES_REQUESTED** [review]
Can you add a test for line deletion and line addition? Seems to display incorrectly when I tested:

image

image

**Comment by KN4CK3R** [comment]
> Can you add a test for line deletion and line addition? Seems to display incorrectly when I tested:

Could you provide a sample file? Works for me as intended.
test

File: web_src/less/_repository.less (hunk 9 above)
Comment (KN4CK3R) [reply web_src/less/_repository.less]: I have removed the line-height. Looks better without it.

**Comment by kdumontnu** [comment]
> > Can you add a test for line deletion and line addition? Seems to display incorrectly when I tested:
> 
> Could you provide a sample file? Works for me as intended.

It worked in other examples, but fails here. Not sure why.

[Tests.zip](https://github.com/go-gitea/gitea/files/6198478/Tests.zip)

**Comment by KN4CK3R** [comment]
The problem is the empty line which is not handled as expected by the Go CSV reader (https://github.com/golang/go/issues/39119). I don't think this PR could and should support all variations in CSV files out there. For usual field,field,field files it works as itended and that should be the majority of CSV files.

**Review by lunny — APPROVED** [review]

**Review by kdumontnu — APPROVED** [review]
Awesome. Thanks for adding unit tests! 💯

**Comment by kdumontnu** [comment]
> The problem is the empty line which is not handled as expected by the Go CSV reader ([golang/go#39119](https://github.com/golang/go/issues/39119)). I don't think this PR could and should support all variations in CSV files out there. For usual field,field,field files it works as itended and that should be the majority of CSV files.

Agreed. I've added some support to the issue/PR you linked, but it appears that the go maintainers are not interested in extending the `encode/csv` library and suggest using other libraries. I might investigate other csv reader libraries.

**Comment by lunny** [comment]
@zeripath need your review.

**Comment by 6543** [comment]
looks ok, just found one dark-theme issue and one render issue:
Bildschirmfoto zu 2021-03-29 01-08-13
Bildschirmfoto zu 2021-03-29 01-08-37

on the Diff View the `®` dont get displayed correctly, while on normal file-view it is ...
both views have an collor issue (white line), whitch should be hidden at all ...

File: modules/csv/csv.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	lines = lines[:util.Min(maxLines, len(lines))]
+
+	delimiters := []rune{',', ';', '\t', '|'}
```
Comment (lafriks) [inline modules/csv/csv.go]: Could you add also `@` as delimiter?

File: modules/csv/csv.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (lafriks) [inline modules/csv/csv.go]:
```suggestion
// Copyright 2021 The Gitea Authors. All rights reserved.
```
please also update year in other newly added files

**Comment by KN4CK3R** [comment]
> on the Diff View the `®` dont get displayed correctly, while on normal file-view it is ...

Could you please provide the sample file? Looks like an encoding issue. I have added a call to `charset.ToUTF8WithFallback` like in `view.go`. There is an unneeded call too (line 485 and 498):
https://github.com/go-gitea/gitea/blob/2b9e0b4d1bfb1b21851e1d56373cbcc52b2c2dfe/routers/repo/view.go#L485-L498

**Review by 6543 — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Sorry accidentally requested review 😸

**Review by kdumontnu — APPROVED** [review]
